extensions/gcp/processors/FetchGCSObject.cpp

```diff
@@ -0,0 +1,196 @@
+  [[nodiscard]] auto getMetaGeneration() const noexcept { return meta_generation_; }
+  [[nodiscard]] auto getStorageClass() const noexcept { return storage_class_; }
+
+
+  void setEncryptionKey(const gcs::EncryptionKey& encryption_key) {
+    encryption_key_ = encryption_key;
+  }
+
+  void setGeneration(const gcs::Generation generation) {
+    generation_ = generation;
+  }
+
+ private:
+  std::string bucket_;
+  std::string key_;
+  gcs::Client& client_;
+
+  gcs::EncryptionKey encryption_key_;
+  gcs::Generation generation_;
+
+  google::cloud::Status status_;
+  std::optional<std::int64_t> result_generation_;
+  std::optional<std::int64_t> meta_generation_;
+  std::optional<std::string> storage_class_;
+};
+}  // namespace
+
+
+void FetchGCSObject::initialize() {
+  setSupportedProperties({GCPCredentials,
+                          Bucket,
+                          Key,
+                          ObjectGeneration,
+                          NumberOfRetries,
+                          EncryptionKey,
+                          EndpointOverrideURL});
+  setSupportedRelationships({Success, Failure});
+}
+
+void FetchGCSObject::onSchedule(const std::shared_ptr<core::ProcessContext>& context, const std::shared_ptr<core::ProcessSessionFactory>& session_factory) {
+  GCSProcessor::onSchedule(context, session_factory);
+  gsl_Expects(context);
+  if (auto encryption_key = context->getProperty(EncryptionKey)) {
+    try {
+      encryption_key_ = gcs::EncryptionKey::FromBase64Key(*encryption_key);
+    } catch (const google::cloud::RuntimeStatusError&) {
+      throw minifi::Exception(ExceptionType::PROCESS_SCHEDULE_EXCEPTION, "Could not decode the base64-encoded encryption key from property " + EncryptionKey.getName());    }
+  }
+}
+
+void FetchGCSObject::onTrigger(const std::shared_ptr<core::ProcessContext>& context, const std::shared_ptr<core::ProcessSession>& session) {
+  gsl_Expects(context && session && gcp_credentials_);
+
+  auto flow_file = session->get();
+  if (!flow_file) {
+    context->yield();
+    return;
+  }
+
+  auto bucket = context->getProperty(Bucket, flow_file);
+  if (!bucket || bucket->empty()) {
+    logger_->log_error("Missing bucket name");
+    session->transfer(flow_file, Failure);
+    return;
+  }
+  auto object_name = context->getProperty(Key, flow_file);
+  if (!object_name || object_name->empty()) {
+    logger_->log_error("Missing object name");
+    session->transfer(flow_file, Failure);
+    return;
+  }
+
+  gcs::Client client = getClient();
+  FetchFromGCSCallback callback(client, *bucket, *object_name);
+  callback.setEncryptionKey(encryption_key_);
+  auto generation_str = context->getProperty(ObjectGeneration, flow_file);
+  if (generation_str) {
+    try {
+      auto generation = std::stol(*generation_str);
+      callback.setGeneration(gcs::Generation(generation));
+    } catch (const std::invalid_argument&) {
+      logger_->log_error("Invalid generation %s", *generation_str);
+    }
```

Comment:
sorry I missed this in the first round, but there should be a transfer to failure + return here, too

Reply:
Replaced the property to use a standard validator instead in https://github.com/apache/nifi-minifi-cpp/pull/1297/commits/90cb3662b891a6626e91bbdedc5c35ed8d5a19db